File: backend/src/main/java/com/agentworkbench/harness/tool/impl/ShellSessionTool.java
package com.agentworkbench.harness.tool.impl;

import com.agentworkbench.harness.core.BackgroundTaskManager;
import com.agentworkbench.harness.safety.PathSandbox;
import com.agentworkbench.harness.shell.OutputManager;
import com.agentworkbench.harness.shell.OutputManager.OutputResult;
import com.agentworkbench.harness.shell.ShellSession;
import com.agentworkbench.harness.shell.ShellSessionManager;
import com.agentworkbench.harness.tool.Tool;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.nio.file.Path;
import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * 唯一的命令执行工具：支持一次性执行和有状态持久会话
 */
@Slf4j
@Component
public class ShellSessionTool implements Tool {

    private static final int MAX_COMMAND_LENGTH = 10000;
    private static final String MARKER_PREFIX = "__CMD_DONE_";
    private static final String MARKER_SUFFIX = "__";

    private final ObjectMapper objectMapper;
    private final PathSandbox pathSandbox;
    private final ShellSessionManager sessionManager;
    private final OutputManager outputManager;
    private final BackgroundTaskManager backgroundTaskManager;

    public ShellSessionTool(ObjectMapper objectMapper, PathSandbox pathSandbox,
                            ShellSessionManager sessionManager, OutputManager outputManager,
                            BackgroundTaskManager backgroundTaskManager) {
        this.objectMapper = objectMapper;
        this.pathSandbox = pathSandbox;
        this.sessionManager = sessionManager;
        this.outputManager = outputManager;
        this.backgroundTaskManager = backgroundTaskManager;
    }

    @Override
    public String getName() {
        return "shell";
    }

    @Override
    public String getDescription() {
        return "执行 shell 命令，支持一次性执行和持久会话。\n" +
                "动作：\n" +
                "- exec：执行命令（如果省略 session_id，则创建新会话）\n" +
                "- write_stdin：向正在运行的会话 stdin 写入输入\n" +
                "- close：关闭 shell 会话\n" +
                "- list：列出活跃会话";
    }

    @Override
    public Map<String, Object> getInputSchema() {
        Map<String, Object> schema = new HashMap<>();
        schema.put("type", "object");

        Map<String, Object> properties = new HashMap<>();

        Map<String, Object> action = new HashMap<>();
        action.put("type", "string");
        action.put("enum", List.of("exec", "write_stdin", "close", "list"));
        action.put("description", "要执行的动作（默认：exec）");
        properties.put("action", action);

        Map<String, Object> command = new HashMap<>();
        command.put("type", "string");
        command.put("description", "要执行的命令（用于 exec 动作）");
        properties.put("command", command);

        Map<String, Object> sessionId = new HashMap<>();
        sessionId.put("type", "string");
        sessionId.put("description", "会话 ID。省略时执行一次性命令；提供时复用已有会话。");
        properties.put("session_id", sessionId);

        Map<String, Object> input = new HashMap<>();
        input.put("type", "string");
        input.put("description", "要写入 stdin 的输入（用于 write_stdin 动作）");
        properties.put("input", input);

        Map<String, Object> workdir = new HashMap<>();
        workdir.put("type", "string");
        workdir.put("description", "工作目录（相对于工作区）");
        properties.put("workdir", workdir);

        Map<String, Object> yieldTimeMs = new HashMap<>();
        yieldTimeMs.put("type", "integer");
        yieldTimeMs.put("description", "等待输出的最长时间，单位毫秒（默认 10000）");
        properties.put("yield_time_ms", yieldTimeMs);

        Map<String, Object> async = new HashMap<>();
        async.put("type", "boolean");
        async.put("description", "是否在后台运行并立即返回 task_id（默认 false，仅用于 exec 动作）");
        properties.put("async", async);

        schema.put("properties", properties);

        return schema;
    }

    @Override
    public Map<String, Object> getOutputSchema() {
        Map<String, Object> schema = new HashMap<>();
        schema.put("type", "object");
        Map<String, Object> properties = new HashMap<>();
        properties.put("exit_code", Map.of("type", "integer"));
        properties.put("session_id", Map.of("type", "string"));
        properties.put("output", Map.of("type", "string"));
        properties.put("current_workdir", Map.of("type", "string"));
        properties.put("truncated", Map.of("type", "boolean"));
        properties.put("async", Map.of("type", "boolean"));
        properties.put("task_id", Map.of("type", "string"));
        schema.put("properties", properties);
        return schema;
    }

    @Override
    public String execute(String arguments) {
        return execute(arguments, null, null);
    }

    @Override
    public String execute(String arguments, String workspace) {
        return execute(arguments, null, workspace);
    }

    @Override
    public String execute(String arguments, Long sessionId, String workspace) {
        try {
            JsonNode args = objectMapper.readTree(arguments);
            String action = args.path("action").asText("exec");
            if (action.isBlank()) {
                action = "exec";
            }

            return switch (action) {
                case "exec" -> handleExec(args, sessionId, workspace);
                case "write_stdin" -> handleWriteStdin(args, sessionId);
                case "close" -> handleClose(args, sessionId);
                case "list" -> handleList(sessionId);
                default -> errorJson("未知动作：" + action);
            };
        } catch (Exception e) {
            log.error("ShellSessionTool execution failed", e);
            return errorJson("错误：" + e.getMessage());
        }
    }

    private String handleExec(JsonNode args, Long conversationId, String workspace) throws Exception {
        String command = args.has("command") ? args.get("command").asText() : null;
        if (command == null || command.isBlank()) {
            return errorJson("exec 动作必须提供 command");
        }
        if (command.length() > MAX_COMMAND_LENGTH) {
            return errorJson("命令过长（最多 " + MAX_COMMAND_LENGTH + " 个字符）");
        }

        String sessionId = args.has("session_id") ? args.get("session_id").asText() : null;
        int yieldTimeMs = args.has("yield_time_ms") ? args.get("yield_time_ms").asInt() : 10000;
        String workdir = args.has("workdir") ? args.get("workdir").asText() : null;
        boolean isAsync = args.has("async") && args.get("async").asBoolean();

        // async 模式：提交后台任务，立即返回
        if (isAsync) {
            String taskId = backgroundTaskManager.submit(() -> {
                try {
                    return doExec(command, sessionId, conversationId, workspace, workdir, yieldTimeMs);
                } catch (Exception e) {
                    return errorJson("异步执行失败：" + e.getMessage());
                }
            });
            return objectMapper.writeValueAsString(Map.of(
                    "async", true,
                    "task_id", taskId,
                    "message", "命令已提交到后台执行。"
            ));
        }

        return doExec(command, sessionId, conversationId, workspace, workdir, yieldTimeMs);
    }

    private String doExec(String command, String sessionId, Long conversationId,
                          String workspace, String workdir, int yieldTimeMs) throws Exception {
        // 获取或创建会话
        ShellSession session = sessionManager.getOrCreate(conversationId, sessionId, workspace);
        sessionId = session.getSessionId();

        // 如果指定了工作目录，先执行 cd
        if (workdir != null) {
            Path resolvedWorkdir = pathSandbox.resolve(workdir, workspace);
            executeWithMarker(session, "cd " + resolvedWorkdir, Duration.ofSeconds(5));
        }

        // 执行命令
        long startTime = System.currentTimeMillis();
        OutputResult output = executeWithMarker(session, command, Duration.ofMillis(yieldTimeMs));
        long elapsedMs = System.currentTimeMillis() - startTime;

        String currentWorkdir = pwdWithMarker(session);

        String result = outputManager.formatToolResult(
                output.markerFound() ? 0 : -1,
                sessionId,
                elapsedMs,
                output,
                currentWorkdir
        );

        session.incrementCommandCount();
        return result;
    }

    private String handleWriteStdin(JsonNode args, Long conversationId) throws Exception {
        String sessionId = args.has("session_id") ? args.get("session_id").asText() : null;
        if (sessionId == null) {
            return errorJson("write_stdin 动作必须提供 session_id");
        }

        String input = args.has("input") ? args.get("input").asText() : null;
        if (input == null) {
            return errorJson("write_stdin 动作必须提供 input");
        }

        int yieldTimeMs = args.has("yield_time_ms") ? args.get("yield_time_ms").asInt() : 5000;

        ShellSession session = sessionManager.getSession(sessionId);
        if (session == null) {
            return errorJson("会话不存在：" + sessionId);
        }

        // 写入输入，用 marker 检测输出结束
        String marker = generateMarker();
        session.getStdin().write(input);
        session.getStdin().newLine();
        session.getStdin().flush();
        // 写入 marker 以便检测输出结束
        session.getStdin().write("echo " + MARKER_PREFIX + marker + MARKER_SUFFIX);
        session.getStdin().newLine();
        session.getStdin().flush();

        Path outputFile = session.nextOutputFile();
        String fullMarker = MARKER_PREFIX + marker + MARKER_SUFFIX;
        OutputResult output = outputManager.readUntilMarker(session.getStdout(), fullMarker,
                Duration.ofMillis(yieldTimeMs), outputFile);
        String currentWorkdir = pwdWithMarker(session);

        session.touch();
        return objectMapper.writeValueAsString(Map.of(
                "session_id", sessionId,
                "current_workdir", currentWorkdir != null ? currentWorkdir : "",
                "output", output.preview(),
                "truncated", output.truncated()
        ));
    }

    private String handleClose(JsonNode args, Long conversationId) throws Exception {
        String sessionId = args.has("session_id") ? args.get("session_id").asText() : null;
        if (sessionId == null) {
            return errorJson("close 动作必须提供 session_id");
        }

        sessionManager.close(sessionId);
        return objectMapper.writeValueAsString(Map.of(
                "session_id", sessionId,
                "status", "已关闭"
        ));
    }

    private String handleList(Long conversationId) throws Exception {
        List<ShellSession> sessions = sessionManager.listByConversation(conversationId);

        List<Map<String, Object>> sessionList = sessions.stream()
                .map(s -> {
                    Map<String, Object> info = new HashMap<>();
                    info.put("session_id", s.getSessionId());
                    info.put("current_workdir", s.getCurrentWorkdir());
                    info.put("command_count", s.getCommandCount());
                    info.put("created_at", s.getCreatedAt().toString());
                    return info;
                })
                .toList();

        return objectMapper.writeValueAsString(Map.of(
                "sessions", sessionList,
                "count", sessionList.size()
        ));
    }

    /**
     * 用 marker 机制执行命令并读取输出
     */
    private OutputResult executeWithMarker(ShellSession session, String command, Duration timeout) {
        String marker = generateMarker();
        String fullMarker = MARKER_PREFIX + marker + MARKER_SUFFIX;

        log.info("executeWithMarker: session={}, command={}, timeoutMs={}, marker={}",
                session.getSessionId(), command, timeout.toMillis(), fullMarker);

        try {
            // 写入命令 + marker echo
            session.getStdin().write(command);
            session.getStdin().newLine();
            session.getStdin().write("echo " + fullMarker);
            session.getStdin().newLine();
            session.getStdin().flush();

            Path outputFile = session.nextOutputFile();
            return outputManager.readUntilMarker(session.getStdout(), fullMarker, timeout, outputFile);
        } catch (Exception e) {
            log.error("Failed to execute command in session {}: {}", session.getSessionId(), e.getMessage());
            return new OutputResult("错误：" + e.getMessage(), 0, 0, false, false, null);
        }
    }

    /**
     * 用 marker 机制获取当前工作目录
     */
    private String pwdWithMarker(ShellSession session) {
        String marker = generateMarker();
        String fullMarker = MARKER_PREFIX + marker + MARKER_SUFFIX;

        try {
            session.getStdin().write("pwd");
            session.getStdin().newLine();
            session.getStdin().write("echo " + fullMarker);
            session.getStdin().newLine();
            session.getStdin().flush();

            OutputResult result = outputManager.readUntilMarker(
                    session.getStdout(), fullMarker, Duration.ofSeconds(5), null);

            String output = result.preview();
            if (output != null && !output.isEmpty()) {
                String[] lines = output.split("\n");
                // pwd 输出在 marker 之前最后一行
                for (int i = lines.length - 1; i >= 0; i--) {
                    String line = lines[i].trim();
                    if (!line.isEmpty() && !line.contains(MARKER_PREFIX)) {
                        session.setCurrentWorkdir(line);
                        return line;
                    }
                }
            }
        } catch (Exception e) {
            log.debug("Failed to get current workdir: {}", e.getMessage());
        }
        return session.getCurrentWorkdir();
    }

    private String generateMarker() {
        return UUID.randomUUID().toString().substring(0, 8);
    }

    private String errorJson(String message) {
        try {
            return objectMapper.writeValueAsString(Map.of(
                    "exit_code", -1,
                    "error", message
            ));
        } catch (Exception e) {
            return "{\"exit_code\":-1,\"error\":\"" + message.replace("\"", "'") + "\"}";
        }
    }
}

File: backend-ts/src/harness/tool/impl/shell-session-tool.ts
import { randomUUID } from 'node:crypto';
import { BaseTool } from '../tool.js';
import { asInt, asText, errorJson, parseObject, toJson } from '../json.js';
import type { PathSandbox } from '../../safety/path-sandbox.js';
import { SecurityException } from '../../safety/path-sandbox.js';
import type { ShellSession, ShellSessionManager, OutputManager, OutputResult } from '../../shell/shell-session-manager.js';
import type { BackgroundTaskManager } from '../../core/background-task-manager.js';
import type { GitCredentialLookup } from '../../../session/types.js';
import { matchShellDenyList } from '../../shell/command-deny-list.js';
import { harnessLog } from '../../log.js';

const MAX_COMMAND_LENGTH = 20000;
const MARKER_PREFIX = '__CMD_DONE_';
const MARKER_SUFFIX = '__';
const WORKDIR_TIMEOUT_MS = 5000;
const DEFAULT_EXEC_YIELD_MS = 300_000;
const DEFAULT_STDIN_YIELD_MS = 5000;
const DEFAULT_AWAIT_YIELD_MS = 60_000;
/** wait_for 由模型给出并在服务端执行匹配，限制长度以压缩灾难性回溯的空间。 */
const MAX_WAIT_FOR_LENGTH = 200;

/** 为 CLOUD shell 签发短效 token 所需的最小依赖。 */
export interface ShellTokenIssuer {
  generateShellToken(userId: number, username: string): string;
}

export interface ShellUserLookup {
  findById(userId: number): Promise<{ username?: string | null } | null>;
}

/** bash 单引号转义，避免 JWT 等特殊字符破坏命令 */
export function shellSingleQuote(value: string): string {
  return `'${value.replace(/'/g, "'\\''")}'`;
}

export class ShellSessionTool extends BaseTool {
  constructor(
    private readonly pathSandbox: PathSandbox,
    private readonly sessionManager: ShellSessionManager,
    private readonly outputManager: OutputManager,
    private readonly backgroundTaskManager: BackgroundTaskManager,
    private readonly gitCredentialService?: GitCredentialLookup | null,
    private readonly jwtService?: ShellTokenIssuer | null,
    private readonly userLookup?: ShellUserLookup | null,
  ) { super(); }

  getName(): string { return 'shell'; }
  getDescription(): string {
    return '执行 shell 命令，支持一次性执行和持久会话。\n'
      + '动作：\n'
      + '- exec：执行命令（如果省略 session_id，则创建新会话）\n'
      + '- write_stdin：向正在运行的会话 stdin 写入输入\n'
      + '- await_async：继续等待会话中未结束的命令（长时构建、后台任务）\n'
      + '- close：关闭 shell 会话\n'
      + '- list：列出活跃会话\n'
      + '参数：\n'
      + '- keep_session：是否保留会话（默认 false），执行后自动关闭会话以释放资源。\n'
      + '- wait_for：正则；命中输出即提前返回（completed=false），命令继续在后台运行。\n'
      + '返回 completed=false 时命令仍在运行，会话被保留，用 await_async + session_id 继续等待；'
      + '完整输出始终写入 output_file。';
  }
  getInputSchema(): Record<string, unknown> {
    return {
      type: 'object',
      properties: {
        action: { type: 'string', enum: ['exec', 'write_stdin', 'await_async', 'close', 'list'], description: '要执行的动作（默认：exec）' },
        command: { type: 'string', description: '要执行的命令（用于 exec 动作）' },
        session_id: { type: 'string', description: '会话 ID。省略时执行一次性命令；提供时复用已有会话。' },
        input: { type: 'string', description: '要写入 stdin 的输入（用于 write_stdin 动作）' },
        workdir: { type: 'string', description: '工作目录：支持相对路径和任意绝对路径' },
        yield_time_ms: { type: 'integer', description: '等待输出的最长时间，单位毫秒（exec 默认 300000，write_stdin 默认 5000，await_async 默认 60000）' },
        wait_for: { type: 'string', description: `正则，命中输出即提前返回（最长 ${MAX_WAIT_FOR_LENGTH} 字符）。例如等服务启动打印 "Listening on"。` },
        async: { type: 'boolean', description: '是否在后台运行并立即返回 task_id（默认 false，仅用于 exec 动作）' },
        task_id: { type: 'string', description: '后台任务 ID（用于 await_async 动作，等待 async 提交的任务）' },
        keep_session: { type: 'boolean', description: '是否保留会话（默认 false）。执行后自动关闭会话以释放资源。' },
      },
    };
  }
  getOutputSchema(): Record<string, unknown> {
    return {
      type: 'object',
      properties: {
        exit_code: { type: 'integer' },
        session_id: { type: 'string' },
        output: { type: 'string' },
        current_workdir: { type: 'string' },
        truncated: { type: 'boolean' },
        completed: { type: 'boolean' },
        matched: { type: 'string' },
        output_file: { type: 'string' },
        async: { type: 'boolean' },
        task_id: { type: 'string' },
      },
    };
  }

  protected async executeWithUser(
    argumentsJson: string, sessionId: number | null, userId: number | null, workspace: string | null,
  ): Promise<string> {
    try {
      const args = parseObject(argumentsJson) ?? {};
      let action = asText(args.action) ?? 'exec';
      if (action.trim() === '') action = 'exec';
      switch (action) {
        case 'exec': return await this.handleExec(args, sessionId, userId, workspace);
        case 'write_stdin': return await this.handleWriteStdin(args, sessionId, userId);
        case 'await_async': return await this.handleAwaitAsync(args, sessionId);
        case 'close': return this.handleClose(args, sessionId);
        case 'list': return this.handleList(sessionId);
        default: return errorJson('未知动作：' + action);
      }
    } catch (e) {
      if (e instanceof SecurityException) harnessLog('warn', `ShellSessionTool blocked by sandbox: ${(e as Error).message}`);
      else harnessLog('error', 'ShellSessionTool execution failed', e);
      return errorJson('错误：' + (e as Error).message);
    }
  }

  /** wait_for 由模型提供，非法正则要作为参数错误反馈而不是抛异常。 */
  private parseWaitFor(args: Record<string, unknown>): RegExp | null | string {
    const raw = asText(args.wait_for);
    if (!raw || raw === '') return null;
    if (raw.length > MAX_WAIT_FOR_LENGTH) return `wait_for 过长（最多 ${MAX_WAIT_FOR_LENGTH} 个字符）`;
    try {
      return new RegExp(raw);
    } catch (e) {
      return `wait_for 不是合法正则：${(e as Error).message}`;
    }
  }

  private async handleExec(
    args: Record<string, unknown>, conversationId: number | null, userId: number | null, workspace: string | null,
  ): Promise<string> {
    const command = asText(args.command);
    if (!command || command.trim() === '') return errorJson('exec 动作必须提供 command');
    if (command.length > MAX_COMMAND_LENGTH) return errorJson(`命令过长（最多 ${MAX_COMMAND_LENGTH} 个字符）`);
    const denied = matchShellDenyList(command);
    if (denied) {
      harnessLog('warn', `Shell command blocked by deny-list [${denied.id}]: ${command}`);
      return errorJson(`命令被拒绝：${denied.reason}`);
    }
    const waitFor = this.parseWaitFor(args);
    if (typeof waitFor === 'string') return errorJson(waitFor);
    const keepSession = args.keep_session === true;
    const isAsync = args.async === true;
    const yieldTimeMs = args.yield_time_ms != null ? asInt(args.yield_time_ms, DEFAULT_EXEC_YIELD_MS) : DEFAULT_EXEC_YIELD_MS;
    let workdir = workspace;
    const workdirArg = asText(args.workdir);
    if (workdirArg) workdir = this.pathSandbox.resolveLenient(workdirArg, workspace);
    const tokenMap = userId != null && this.gitCredentialService
      ? await this.gitCredentialService.getTokenMapByUser(userId) : {};
    const shellId = asText(args.session_id);
    const session = this.sessionManager.getOrCreate(conversationId ?? 0, shellId, userId, workdir, tokenMap);
    const releaseCommand = await session.acquireCommand?.() ?? (() => undefined);
    // 上一条命令未结束时再写新命令会让两条命令的输出交织，且新命令要排在它之后才执行
    if (session.pendingCommand) {
      releaseCommand();
      return errorJson(`会话仍有未结束的命令：${session.sessionId}。请先用 action:'await_async' 收取结果，或 close 该会话。`);
    }
    try {
      // 持久会话的用户环境不会自动更新，执行前按当前触发者刷新 Git/Home 环境。
      this.sessionManager.refreshUserEnvironment?.(session, userId, tokenMap);
      await this.injectMaoToken(session, userId);
      // 复用已有会话时 getOrCreate 不会改变 cwd，必须显式 cd
      if (workdirArg && workdir) {
        await this.executeWithMarker(session, 'cd ' + shellSingleQuote(workdir), WORKDIR_TIMEOUT_MS);
      }
      const marker = this.newMarker();
      if (isAsync) {
        let taskId: string;
        try {
          taskId = this.backgroundTaskManager.submit(conversationId, async () => {
            try {
              const r = await this.outputManager.readUntilMarker(session, marker, yieldTimeMs, waitFor);
              this.settleSession(session, r, keepSession);
              return toJson(this.formatResult(session, r, session.currentWorkdir));
            } finally {
              // 后台任务退出后读取权交还会话，后续 await_async 才能接着读
              if (session.pendingCommand) session.pendingCommand.taskId = null;
              releaseCommand();
            }
          });
        } catch (error) {
          releaseCommand();
          throw error;
        }
        this.writeCommand(session, command, marker, keepSession, taskId);
        return toJson({
          async: true,
          task_id: taskId,
          session_id: session.sessionId,
          output_file: session.outputFile,
          message: '命令已提交到后台执行。',
        });
      }
      this.writeCommand(session, command, marker, keepSession, null);
      const result = await this.outputManager.readUntilMarker(session, marker, yieldTimeMs, waitFor);
      const workdirNow = await this.resolveCurrentWorkdir(session, result);
      const payload = this.formatResult(session, result, workdirNow);
      this.settleSession(session, result, keepSession);
      return toJson(payload);
    } finally {
      if (!isAsync) releaseCommand();
    }
  }

  /** 写入命令并登记为等待中；提前返回后仍能凭 marker 继续读。 */
  private writeCommand(
    session: ShellSession, command: string, marker: string, keepSession: boolean, taskId: string | null,
  ): void {
    session.beginCommand(marker, keepSession, true, taskId);
    session.writeStdin(command + '\necho ' + marker + ' $?\n');
    session.incrementCommandCount();
    session.touch();
  }

  /**
   * 命令已结束才按 keep_session 决定是否回收；仍在运行时必须保留会话，
   * 否则关闭会 SIGKILL 掉进程组，模型再也拿不到剩余输出。
   */
  private settleSession(session: ShellSession, result: OutputResult, keepSession: boolean): void {
    if (result.completed && !keepSession) this.sessionManager.close(session.sessionId);
  }

  private formatResult(
    session: ShellSession, result: OutputResult, currentWorkdir: string,
  ): Record<string, unknown> {
    const payload: Record<string, unknown> = {
      exit_code: this.resolveExitCode(result),
      session_id: session.sessionId,
      output: result.output,
      truncated: result.truncated,
      completed: result.completed,
      current_workdir: currentWorkdir,
      output_file: session.outputFile,
    };
    if (result.matched != null) payload.matched = result.matched;
    if (!result.completed) {
      payload.message = result.matched != null
        ? `wait_for 已命中，命令仍在运行。用 action:'await_async' + session_id:'${session.sessionId}' 继续等待。`
        : `等待超时，命令仍在运行。用 action:'await_async' + session_id:'${session.sessionId}' 继续等待。`;
    }
    return payload;
  }

  private async handleWriteStdin(args: Record<string, unknown>, conversationId: number | null, userId: number | null): Promise<string> {
    const shellId = asText(args.session_id);
    const input = asText(args.input) ?? '';
    if (!shellId) return errorJson('write_stdin 必须提供 session_id');
    const denied = matchShellDenyList(input);
    if (denied) {
      harnessLog('warn', `Shell stdin blocked by deny-list [${denied.id}]: ${input}`);
      return errorJson(`命令被拒绝：${denied.reason}`);
    }
    const waitFor = this.parseWaitFor(args);
    if (typeof waitFor === 'string') return errorJson(waitFor);
    const session = this.sessionManager.getSession(shellId);
    if (!session) return errorJson('会话不存在或已关闭：' + shellId);
    const pending = session.pendingCommand;
    if (pending?.taskId != null) {
      return errorJson(`会话的输出正由后台任务 ${pending.taskId} 读取，请先 await_async 该任务。`);
    }
    const releaseCommand = await session.acquireCommand?.() ?? (() => undefined);
    try {
      const yieldTimeMs = args.yield_time_ms != null ? asInt(args.yield_time_ms, DEFAULT_STDIN_YIELD_MS) : DEFAULT_STDIN_YIELD_MS;
      // 拿锁可能排队（后台任务/并行调用持有锁），必须以拿锁后的实时状态为准，进锁前的快照可能已过期
      const livePending = session.pendingCommand;
      if (livePending) {
        if (session.peekBuffer().includes(livePending.marker)) {
          // 命令已结束但结果尚未被收取：此时写入的输入会被 bash 当作新命令执行，
          // 而续读会立刻命中旧 marker，把旧输出误当成输入的应答
          return errorJson(`上一条命令已结束但结果尚未收取：${shellId}。请先 action:'await_async' 收取结果，再发送新输入。`);
        }
        // 有命令正在运行：输入交给它，不能再插入 marker（marker 只会排在该命令之后被执行）
        session.writeStdin(input.endsWith('\n') ? input : input + '\n');
        session.touch();
        const answered = await this.outputManager.readUntilMarker(session, livePending.marker, yieldTimeMs, waitFor);
        const workdirNow = await this.resolveCurrentWorkdir(session, answered);
        const payload = this.formatResult(session, answered, workdirNow);
        this.settleSession(session, answered, livePending.keepSession);
        return toJson(payload);
      }
      const tokenMap = userId != null && this.gitCredentialService
        ? await this.gitCredentialService.getTokenMapByUser(userId) : {};
      this.sessionManager.refreshUserEnvironment?.(session, userId, tokenMap);
      // 与 exec 一致：写命令前重新注入短效 MAO_TOKEN
      await this.injectMaoToken(session, userId);
      // 输入本身不带结束标记，必须额外回显 marker，否则只能空等到超时
      const marker = this.newMarker();
      this.writeCommand(session, input, marker, true, null);
      const result = await this.outputManager.readUntilMarker(session, marker, yieldTimeMs, waitFor);
      return toJson(this.formatResult(session, result, await this.resolveCurrentWorkdir(session, result)));
    } finally {
      releaseCommand();
    }
  }

  /** 继续等待未结束的命令：按 task_id 等后台任务，或按 session_id 直接续读会话。 */
  private async handleAwaitAsync(args: Record<string, unknown>, conversationId: number | null): Promise<string> {
    const waitFor = this.parseWaitFor(args);
    if (typeof waitFor === 'string') return errorJson(waitFor);
    const yieldTimeMs = args.yield_time_ms != null ? asInt(args.yield_time_ms, DEFAULT_AWAIT_YIELD_MS) : DEFAULT_AWAIT_YIELD_MS;
    const taskId = asText(args.task_id);
    if (taskId) {
      const awaited = await this.backgroundTaskManager.awaitResult(taskId, yieldTimeMs, conversationId);
      if (awaited.status === 'not_found') return errorJson('后台任务不存在或已被消费：' + taskId);
      if (awaited.status === 'pending') {
        return toJson({
          task_id: taskId,
          completed: false,
          message: `后台任务仍在执行（已等待 ${yieldTimeMs} ms），可继续 await_async 或等待结果自动注入。`,
        });
      }
      return awaited.result;
    }
    const shellId = asText(args.session_id);
    if (!shellId) return errorJson("await_async 必须提供 session_id 或 task_id");
    const session = this.sessionManager.getSession(shellId);
    if (!session) return errorJson('会话不存在或已关闭：' + shellId);
    const pending = session.pendingCommand;
    if (!pending) return errorJson('该会话没有未结束的命令：' + shellId);
    if (pending.taskId != null) {
      return errorJson(`会话的输出正由后台任务 ${pending.taskId} 读取，请用 task_id:'${pending.taskId}' 等待。`);
    }
    // 并行工具调用下两个 await_async 同时轮询同一缓冲区会互相吃掉输出，这里串行化
    const releaseCommand = await session.acquireCommand?.() ?? (() => undefined);
    try {
      // 拿锁可能排队，期间命令可能已被并行调用方收取完毕；按旧快照续读只会登记假 marker 卡死会话
      const livePending = session.pendingCommand;
      if (!livePending) return errorJson('该会话没有未结束的命令：' + shellId);
      const result = await this.outputManager.readUntilMarker(session, livePending.marker, yieldTimeMs, waitFor);
      const workdirNow = await this.resolveCurrentWorkdir(session, result);
      const payload = this.formatResult(session, result, workdirNow);
      this.settleSession(session, result, pending.keepSession);
      return toJson(payload);
    } finally {
      releaseCommand();
    }
  }

  private resolveExitCode(result: OutputResult): number {
    if (result.exitCode != null) return result.exitCode;
    return result.completed ? 0 : -1;
  }

  /**
   * 命令仍在运行时不能再向同一 shell 写 pwd：它会排在当前命令之后，
   * 读取过程反而会消费当前命令的后续输出并把日志末行误判为工作目录。
   */
  private async resolveCurrentWorkdir(session: ShellSession, result: OutputResult): Promise<string> {
    if (!result.completed) return session.currentWorkdir;
    try {
      const marker = this.newMarker();
      // pwd 属协议命令，输出不进 output_file
      session.beginCommand(marker, true, false);
      session.writeStdin('pwd\necho ' + marker + '\n');
      const pwd = await this.outputManager.readUntilMarker(session, marker, WORKDIR_TIMEOUT_MS);
      if (pwd.completed) {
        const lines = pwd.output.split('\n').map((l) => l.trim()).filter((l) => l !== '');
        const last = lines[lines.length - 1];
        if (last && last.startsWith('/')) {
          session.setCurrentWorkdir(last);
          return last;
        }
      }
    } catch (e) {
      harnessLog('debug', `Failed to refresh shell workdir: ${(e as Error).message}`);
    }
    return session.currentWorkdir;
  }

  private async executeWithMarker(session: ShellSession, command: string, timeoutMs: number): Promise<OutputResult> {
    const marker = this.newMarker();
    session.beginCommand(marker, true, false);
    session.writeStdin(command + '\necho ' + marker + ' $?\n');
    return this.outputManager.readUntilMarker(session, marker, timeoutMs);
  }

  /** 为 CLOUD shell 注入短效 JWT，供 mao-*-cli 使用。必须在写入命令前完成，否则 export 会排到命令之后。 */
  private async injectMaoToken(session: ShellSession, userId: number | null): Promise<void> {
    if (userId == null || !this.jwtService || !this.userLookup) return;
    try {
      const user = await this.userLookup.findById(userId);
      const username = user?.username;
      if (!username || username.trim() === '') {
        harnessLog('warn', `Skip MAO_TOKEN inject: user not found or username empty, userId=${userId}`);
        return;
      }
      const token = this.jwtService.generateShellToken(userId, username);
      session.writeStdin('export MAO_TOKEN=' + shellSingleQuote(token) + '\n');
    } catch (e) {
      harnessLog('warn', `Failed to inject MAO_TOKEN for userId=${userId}: ${(e as Error).message}`);
    }
  }

  private newMarker(): string {
    return MARKER_PREFIX + randomUUID().replace(/-/g, '').slice(0, 12) + MARKER_SUFFIX;
  }

  private handleClose(args: Record<string, unknown>, _conversationId: number | null): string {
    const shellId = asText(args.session_id);
    if (!shellId) return errorJson('close 必须提供 session_id');
    this.sessionManager.close(shellId);
    return toJson({ success: true, session_id: shellId });
  }

  private handleList(conversationId: number | null): string {
    const sessions = conversationId != null ? this.sessionManager.listByConversation(conversationId) : [];
    return toJson({
      sessions: sessions.map((s) => ({
        session_id: s.sessionId,
        current_workdir: s.currentWorkdir,
        alive: s.isAlive(),
      })),
    });
  }
}
